src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp

```diff
@@ -0,0 +1,492 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "appleseedrenderelement.h"
+
+// appleseed-max headers.
+#include "appleseedrenderelement/resource.h"
+#include "appleseedrenderer/appleseedrenderer.h"
+#include "main.h"
+#include "utilities.h"
+#include "version.h"
+
+// appleseed.renderer headers.
+#include "renderer/api/aov.h"
+
+// 3ds Max headers.
+#include <iparamm2.h>
+
+// Standard headers.
+#include <string>
+#include <vector>
+
+namespace asf = foundation;
+namespace asr = renderer;
+
+namespace
+{
+    const wchar_t* AppleseedRenderElementFriendlyClassName = L"appleseed_AOV";
+
+    enum { ParamBlockIdRenderElement };
+    enum { ParamBlockRefRenderElement };
+
+    enum ParamId
+    {
+        ParamIdEnableOn         = 0,
+        ParamIdElementBitmap    = 1,
+        ParamIdElementName      = 2,
+        ParamIdAOVIndex         = 3,
+    };
+}
+
+AppleseedRenderElementClassDesc g_appleseed_renderelement_classdesc;
+asr::AOVFactoryRegistrar g_aov_factory_registrar;
+
+
+//
+// AppleseedRenderElement class implementation.
+//
+
+namespace
+{
+    
+    std::vector<std::string> get_aov_names()
+    {
+        auto factories = asf::array_vector<std::vector<asr::IAOVFactory*>>(g_aov_factory_registrar.get_factories());
+        std::vector<std::string> aov_names;
+        for (const auto& factory : factories)
+            aov_names.push_back(factory->get_model_metadata().get("label"));
+
+        return aov_names;
+    }
+
+    void init_combo_box(IParamMap2* param_map)
+    {
+        HWND dlg_hwnd = nullptr;
+        if (param_map != nullptr)
+            dlg_hwnd = param_map->GetHWnd();
+        if (dlg_hwnd == nullptr)
+            return;
+
+        SendMessage(GetDlgItem(dlg_hwnd, IDC_COMBO_AOV_NAME), CB_RESETCONTENT, 0, 0);
+        if (true)
+        {
+            auto aov_names = get_aov_names();
+
+            for (const auto& name : aov_names)
+            {
+                SendMessage(
+                    GetDlgItem(dlg_hwnd, IDC_COMBO_AOV_NAME),
+                    CB_ADDSTRING,
+                    0,
+                    reinterpret_cast<LPARAM>(utf8_to_wide(name).c_str()));
+            }
+
+            int output_index = param_map->GetParamBlock()->GetInt(ParamIdAOVIndex, 0, FOREVER);
+            DbgAssert(output_index >= 1);
+            output_index -= 1;
+            if (output_index < aov_names.size())
+                SendMessage(GetDlgItem(dlg_hwnd, IDC_COMBO_AOV_NAME), CB_SETCURSEL, output_index, 0);
+            else
+                SendMessage(GetDlgItem(dlg_hwnd, IDC_COMBO_AOV_NAME), CB_SETCURSEL, -1, 0);
+        }
+    }
+
+    class AppleseedRenderElementParamDlgProc
+      : public ParamMap2UserDlgProc
+    {
+      public:
+        void DeleteThis() override
+        {
```

Comment:
Are we sure not doing anything here is correct?

Reply:
I thought it's fine since instance of this class is only created as a local object in anonymous namespace (line 176).

Reply:
You're right, it's not an actual bug, however it's better to make the code as agnostic as possible to specific usage patterns, hence I suggest implementing this method (just `delete this` I presume?).

Reply:
I think usage pattern should be change in this case, otherwise I can't open rendering dialog box, it crashes the max.
I'll add `delete this` and assign dialog procedure in a way it is done in other parts of the plugin.